agent/proxycfg/state.go

```diff
@@ -0,0 +1,357 @@
+			if !snap.Valid() {
+				// Not valid yet just respond with nil and move on to next task.
+				replyCh <- nil
+				continue
+			}
+			// Make a deep copy of snap so we don't mutate any of the embedded structs
+			// etc on future updates.
+			snapCopy, err := snap.Clone()
+			if err != nil {
+				s.logger.Printf("[ERR] Failed to copy config snapshot for proxy %s",
+					s.proxyID)
+				continue
+			}
+			replyCh <- snapCopy
+
+			// Skip rest of loop - there is nothing to send since nothing changed on
+			// this iteration
+			continue
+		}
+
+		// Check if snap is complete enough to be a valid config to deliver to a
+		// proxy yet.
+		if snap.Valid() {
+			// Don't send it right away, set a short timer that will wait for updates
+			// from any of the other cache values and deliver them all together.
+			if coallesceTimer == nil {
+				coallesceTimer = time.AfterFunc(coallesceTimeout, func() {
+					// This runs in another goroutine so we can't just do the send
+					// directly here as access to snap is racy so just signal the main
+					// loop above.
+					sendCh <- struct{}{}
+				})
+			}
+		}
+	}
+}
+
+func (s *State) handleUpdate(u cache.UpdateEvent, snap *ConfigSnapshot) error {
+	switch u.CorrelationID {
+	case "roots":
+		roots, ok := u.Result.(*structs.IndexedCARoots)
+		if !ok {
+			return fmt.Errorf("invalid type for roots response: %T", u.Result)
+		}
+		snap.Roots = roots
+	case "leaf":
+		leaf, ok := u.Result.(*structs.IssuedCert)
+		if !ok {
+			return fmt.Errorf("invalid type for leaf response: %T", u.Result)
+		}
+		snap.Leaf = leaf
+	case "intentions":
+		// Not in snapshot currently, no op
+	default:
+		// Service discovery result, figure out which type
+		if strings.HasPrefix(u.CorrelationID, "service:") {
+			resp, ok := u.Result.(*structs.IndexedCheckServiceNodes)
+			if !ok {
+				return fmt.Errorf("invalid type for service response: %T", u.Result)
+			}
+			snap.UpstreamEndpoints[u.CorrelationID] = resp.Nodes
+		} else if strings.HasPrefix(u.CorrelationID, "prepared_query:") {
+			resp, ok := u.Result.(*structs.PreparedQueryExecuteResponse)
+			if !ok {
+				return fmt.Errorf("invalid type for prepared query response: %T", u.Result)
+			}
+			snap.UpstreamEndpoints[u.CorrelationID] = resp.Nodes
+		} else {
+			return errors.New("unknown correlation ID")
+		}
+	}
+	return nil
+}
+
+// CurrentSnapshot synchronously returns the current ConfigSnapshot if there is
+// one ready. If we don't have one yet because not all necessary parts have been
+// returned (i.e. both roots and leaf cert), nil is returned.
+func (s *State) CurrentSnapshot() *ConfigSnapshot {
+	// Make a chan for the response to be sent on
+	ch := make(chan *ConfigSnapshot, 1)
+	s.reqCh <- ch
+	// Wait for the response
+	return <-ch
+}
+
+// Close discards the state and stops any long-running watches.
+func (s *State) Close() error {
+	if s.cancel != nil {
+		s.cancel()
+		s.cancel = nil
+	}
+	if s.snapCh != nil {
+		close(s.snapCh)
+		s.snapCh = nil
```

Comment:
Unfortunately this appears to be a read/write race the running goroutine.

The cleanest way I've found to do this is to pass the channel in as an arg to the unexported `run`. That way it doesn't need to access shared struct state at all, its all local. But, you also return the channel from `Watch` so it looks like you need a lock regardless.

Reply:
Yeah I'll refactor this. Was a rushed attempt to make Close idempotent. I think it's cleaner to just use something else like an atomic int flag to avoid the double close rather than making the members nil which is prone to races.

Reply:
Even worse there is a panic here - we close this chan to signal to manager to end the broadcasting goroutine but we are also sending to it in `run` which means especially bad timing could cause a panic if we try to send a new snapshot just after this closes but before `run` notices the context was cancelled.

I'm going to refactor and simplify all of this a bit I think.